### PR TITLE
Add Stack<T> converter example

### LINKED ISSUE #16690: Add Stack<T> converter example
Add example converter that serializes Stack objects without reversing the order.

https://github.com/dotnet/docs/pull/16225#discussion_r366715884

Start the new section with text that was in the Migrate doc:

```
In <xref:System.Text.Json>, the order of a stack's contents is reversed when it's serialized.
This behavior applies to the following types and interface and user-defined types that
derive from them:	
* <xref:System.Collections.Stack>	
* <xref:System.Collections.Generic.Stack%601>	
* <xref:System.Collections.Immutable.ImmutableStack%601>	
* <xref:System.Collections.Immutable.IImmutableStack%601>	
A custom converter could be implemented to keep stack contents in the same order.
```
---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 5de37d79-751f-e372-e866-efde28f4908c
* Version Independent ID: 678c4f23-764a-7ccd-0d6d-159a30957477
* Content: [How to write custom converters for JSON serialization - .NET](https://docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-converters-how-to#feedback)
* Content Source: [docs/standard/serialization/system-text-json-converters-how-to.md](https://github.com/dotnet/docs/blob/master/docs/standard/serialization/system-text-json-converters-how-to.md)
* Product: **dotnet**
* GitHub Login: @tdykstra
* Microsoft Alias: **tdykstra**

### PR DESCRIPTION
**Pull request by tdykstra**
Fixes #16690 

cc @ahsonkhan

### CODE REVIEW

**Comment by steveharter** [comment]
@layomia do we have an issue to round-trip stack "properly" for 5.0?

**Review by BillWagner — APPROVED** [review]
This LGTM @tdykstra. You can :shipit: when ready.

When would you want to move these snippets into the new location?

**Comment by tdykstra** [comment]
> When would you want to move these snippets into the new location?

I think the diff  would be too confusing to do it in the same PR as an update to the code, so I'll do it separately, will try to get to it this week.

**Comment by layomia** [comment]
> @layomia do we have an issue to round-trip stack "properly" for 5.0?

We did, but closed it because there's no standard way to fix the bug. Our suggested way (as shown in the doc example in this PR) is to reverse the items on serialization, but one could argue that the items should be reversed on deserialization - https://github.com/dotnet/runtime/issues/31211#issuecomment-603886279. We can reexamine the decision if we feel strongly that our suggested way is correct/adequate for most scenarios.

File: docs/standard/serialization/system-text-json-converters-how-to.md [hunk 1]
```diff
@@ -279,6 +281,18 @@ The converter can deserialize JSON that was created by using the same converter
 
 The converter code in the preceding example reads and writes each property manually. An alternative is to call `Deserialize` or `Serialize` to do some of the work. For an example, see [this StackOverflow post](https://stackoverflow.com/a/59744873/12509023).
 
+### Support round-trip for Stack\<T>
+
+If you deserialize a JSON string into a <xref:System.Collections.Generic.Stack%601> object and then serialize that object, the contents of the stack are in reverse order. To support serialization and deserialization that retains the original order in the stack, a custom converter is required.
+
+The following code shows a custom converter that enables round trips to and from `Stack<T>` objects:
```
Comment (ahsonkhan) [inline docs/standard/serialization/system-text-json-converters-how-to.md:288]:
```suggestion
The following code shows a custom converter that enables round-tripping to and from `Stack<T>` objects:
```

File: samples/snippets/core/system-text-json/csharp/RoundtripStackOfT.cs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+using System.Collections.Immutable;
+using System.Text.Json;
+
+namespace SystemTextJsonSamples
+{
+    public class RoundtripStackOfT
+    {
+        public static void Run()
+        {
+            Console.WriteLine("Deserialize JSON string [1, 2, 3], then deserialize");
```
Comment (ahsonkhan) [inline samples/snippets/core/system-text-json/csharp/RoundtripStackOfT.cs:13]:
```suggestion
            Console.WriteLine("Deserialize JSON string [1, 2, 3], then serialize it back to JSON.");
```

File: samples/snippets/core/system-text-json/csharp/RoundtripStackOfT.cs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+            Stack<int> stack = JsonSerializer.Deserialize<Stack<int>>("[1, 2, 3]");
+            string serialized = JsonSerializer.Serialize(stack);
+            Console.WriteLine($"Result is reverse order {serialized}");
```
Comment (ahsonkhan) [inline samples/snippets/core/system-text-json/csharp/RoundtripStackOfT.cs:16]:
```suggestion
            Console.WriteLine($"Result is in reverse order: {serialized}");
```

File: samples/snippets/core/system-text-json/csharp/RoundtripStackOfT.cs [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+            Console.WriteLine("Deserialize JSON string [1, 2, 3] with custom converter, then deserialize");
```
Comment (ahsonkhan) [inline samples/snippets/core/system-text-json/csharp/RoundtripStackOfT.cs:18]:
```suggestion
            Console.WriteLine("Deserialize JSON string [1, 2, 3] with custom converter, then serialize it back to JSON.");
```

File: samples/snippets/core/system-text-json/csharp/RoundtripStackOfT.cs [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+            // <SnippetRegister>
+            var options = new JsonSerializerOptions
+            {
+                Converters = { new JsonConverterFactoryForStackOfT() },
+            };
+            // </SnippetRegister>
+            stack = JsonSerializer.Deserialize<Stack<int>>("[1, 2, 3]", options);
+            serialized = JsonSerializer.Serialize(stack, options);
+            Console.WriteLine($"Result is same order {serialized}");
```
Comment (ahsonkhan) [inline samples/snippets/core/system-text-json/csharp/RoundtripStackOfT.cs:27]:
```suggestion
            Console.WriteLine($"Result is in same order: {serialized}");
```

**Review by ahsonkhan — APPROVED** [review]
Otherwise, looks good.

File: docs/standard/serialization/system-text-json-converters-how-to.md [hunk 6]
```diff
@@ -279,6 +281,18 @@ The converter can deserialize JSON that was created by using the same converter
 
 The converter code in the preceding example reads and writes each property manually. An alternative is to call `Deserialize` or `Serialize` to do some of the work. For an example, see [this StackOverflow post](https://stackoverflow.com/a/59744873/12509023).
 
+### Support round-trip for Stack\<T>
+
+If you deserialize a JSON string into a <xref:System.Collections.Generic.Stack%601> object and then serialize that object, the contents of the stack are in reverse order. To support serialization and deserialization that retains the original order in the stack, a custom converter is required.
```
Comment (ahsonkhan) [inline docs/standard/serialization/system-text-json-converters-how-to.md:286]: From the issue description (https://github.com/dotnet/docs/issues/16690), should we mention all the types that this impacts (including types that derive from them)?

@layomia, is this list still accurate?
```text
This behavior applies to the following types and interface and user-defined types that
derive from them:	
* <xref:System.Collections.Stack>	
* <xref:System.Collections.Generic.Stack%601>	
* <xref:System.Collections.Immutable.ImmutableStack%601>	
* <xref:System.Collections.Immutable.IImmutableStack%601>	
```

Comment (layomia) [reply docs/standard/serialization/system-text-json-converters-how-to.md]: also `ConcurrentStack<T>`.

**Pull request merged**